### PR TITLE
[BugFix] fix iceberg partition transformed issues

### PR DESCRIPTION
**Pull request by stephen-shelby**
Fixes #issue
https://github.com/StarRocks/starrocks/issues/33714

TODO: Use the new `toHumanString` api in next time refactor

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by dirtysalt — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java [hunk 1]
```diff
@@ -314,7 +315,32 @@ public List<PartitionKey> getPrunedPartitions(Table table, ScalarOperator predic
...
+                for (PartitionField partitionField : spec.fields()) {
+                    if (partitionField.transform().isVoid()) {
+                        continue;
+                    }
+
+                    if (!partitionField.transform().isIdentity()) {
+                        Type sourceType = spec.schema().findType(partitionField.sourceId());
+                        Type resultType = partitionField.transform().getResultType(sourceType);
+                        if (resultType == Types.DateType.get()) {
+                            resultType = Types.IntegerType.get();
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java:328]: why transform date type to integer

Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java:328]: the partition value of data type is Integer.

File: fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java [hunk 2]
```diff
@@ -201,6 +198,11 @@ public List<Integer> getSortKeyIndexes() {
         return indexes;
     }
 
+    // day(dt) -> identity dt
+    public boolean hasPartitionTransformedEvolution() {
+        return getNativeTable().spec().fields().stream().anyMatch(field -> field.transform().isVoid());
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java:203]: Does this mean it use a expr parition?

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java [hunk 3]
```diff
@@ -314,7 +315,32 @@ public List<PartitionKey> getPrunedPartitions(Table table, ScalarOperator predic
...
+                        }
+                        srTypes.add(fromIcebergType(resultType));
+                        continue;
+                    }
+
+                    srTypes.add(icebergTable.getColumn(partitionField.name()).getType());
+                }
+
+                if (icebergTable.hasPartitionTransformedEvolution()) {
+                    srTypes = partitionColumns.stream()
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java:338]: Will srTypes be the transformed type after expr?

File: fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java (hunk 2 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/catalog/IcebergTable.java:203]: it means iceberg table has alter `partition pattern` on transformed partitioned column. for explame:  `dt -> day(dt)` or `day(dt) -> dt` or `month(dt) -> day(dt) `.

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java (hunk 3 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/IcebergMetadata.java:338]: This type is only used to create partitionkey

**Review by Youngwb — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Pull request merged**